airbyte-cdk/java/airbyte-cdk/db-sources/src/main/kotlin/io/airbyte/cdk/components/ComponentRunner.kt

```diff
@@ -0,0 +1,131 @@
+                log.warn("$name sentinel threw an error or an exception", e)
+            }
+        }
+        val sentinelThread = Thread(sentinel, "$name-closer-sentinel")
+        val producer: ProducerComponent<S> =
+            producerBuilder.build(input, consumer, sentinelThread::interrupt)
+        val producerThread = Thread(producer, "$name-producer")
+        val closer = Runnable {
+            try {
+                sentinelThread.join()
+                log.debug("$name timed out, closing...")
+            } catch (e: InterruptedException) {
+                log.debug("$name was notified to stop before timing out, closing...", e)
+            }
+            try {
+                producer.close()
+                log.debug("$name closed")
+            } catch (e: Throwable) {
+                log.warn("$name producer close() threw an exception", e)
+                throw e
+            }
+        }
+        val closerThread = Thread(closer, "$name-closer")
+        val thrown = AtomicReference<Throwable>()
+        log.debug("$name is running")
+        sentinelThread.start()
+        producerThread.setUncaughtExceptionHandler { _, e: Throwable ->
+            thrown.set(e)
+            sentinelThread.interrupt()
+            closerThread.interrupt()
+            log.warn("$name producer run() threw an exception, interrupted other $name threads", e)
+        }
+        producerThread.start()
+        closerThread.start()
+        try {
+            closerThread.join()
+            producerThread.join()
+            sentinelThread.join()
+        } catch (e: InterruptedException) {
+            log.debug("$name was interrupted", e)
+        }
+        thrown.get()?.run { throw this }
+        log.debug("$name is done")
+        return Pair(consumer.flush(), producer.finalState())
+    }
+
+    /**
+     * Calls [collect] repeatedly until:
+     * - either the state has reached or exceeded its upper bound,
+     * - or there is no more forward progress from one state to the next.
+     *
+     * It does this lazily.
+     */
+    fun collectRepeatedly(initialState: S, upperBound: S): Sequence<Pair<Sequence<R>, S>> =
+        Sequence {
+            object : Iterator<Pair<Sequence<R>, S>> {
+
+                var nextValue: Pair<Sequence<R>, S>? = collect(initialState)
+
+                override fun hasNext(): Boolean = nextValue != null
+
+                override fun next(): Pair<Sequence<R>, S> {
+                    val currentValue = nextValue ?: throw NoSuchElementException()
+                    val (_, currentState) = currentValue
+                    val hasReachedUpperBound = comparator.compare(currentState, upperBound) >= 0
+                    if (hasReachedUpperBound) {
+                        nextValue = null
+                    } else {
+                        // Compute the successor output ahead of time for the next iteration.
+                        val (successorOutput, successorState) = collect(currentState)
```

Comment:
How do we deal with the case where the stopping criteria for `collect` is reached, but the debezium state hasn't been updated? 

I believe there are cases with records assc with a large transaction where we can't always guarantee the debezium state will progress over a fixed number of records